### DISCUSSION TITLE
how to render a face with diffuse_uv_map, specular_uv_map and roughness_uv_map?

### DISCUSSION THREAD

**Discussion by anewusername77** [mitsuba-renderer/mitsuba3#274] [Q&A] [answered]
Thanks.

**Comment by Speierers** [comment, marked as answer]
@anewusername77 please take the time to better describe what it is you are trying to achieve, what you have already tried, what failed and what succeeded. Then we will be able to answer your questions.

**Reply by anewusername77** [reply]
Thanks so much for your reply.
I'm new to mitsuba,  following your [tutorials](https://mitsuba.readthedocs.io/en/latest/src/quickstart/mitsuba_quickstart.html) and testing by change the .xml scene file.
I got 3 texture maps of a face as 3 channel pictures, which are diffuse reflectance map, specular reflectance map, and roughness reflectance map, also one environment map and a mesh obj file. Now, I want to render this face with Cook-Torrance BRDF model, so I tried [mitsuba BRDFs the principled](https://mitsuba.readthedocs.io/en/latest/src/generated/plugins_bsdfs.html?highlight=principled#the-principled-bsdf-principled), but the result seemed that specular texture wasn't there, couldn't be seen from the man's eyes or skin, quiet far from the result I got from [pyredner](https://github.com/BachiLi/redner/tree/master/pyredner), which also uses ray tracing to render(but much slower).

cause I'm not allowed to upload anything from my lab PC, you can see texture maps similar as mine in [paper figure 2](https://arxiv.org/pdf/2103.15432.pdf), sorry for taking your time. Big thanks!

**Reply by Speierers** [reply]
Using the `principled` BSDF is probably the right thing to do here. Maybe it is the illumination in your scene that is off somehow?

**Reply by anewusername77** [reply]
thanks a lot, it helps. 
Two more questions: 1. should I put specular texture map in `spec tint`?  2. how can I use spherical harmonics to render this face? should I just put the env_map version of  spherical harmonics in the scene file?

**Reply by giuvecchio** [reply]
Any updates? I'm facing the same issue with using a specular map with ```principled``` BSDF

**Reply by njroussel** [reply]
What is your exact issue?
The `specular` and `spec_tint` parameter should be what you are looking for. The two papers that are referenced in the documentation of the `principled` BSDF might also be helpful to you to understand the model.

**Reply by anewusername77** [reply]
> What is your exact issue? The `specular` and `spec_tint` parameter should be what you are looking for. The two papers that are referenced in the documentation of the `principled` BSDF might also be helpful to you to understand the model.

hi njroussel,
I managed to solve it! by using `mi.traverse()`to change `params['vertex_texcoords']` with the help of this doc: https://mitsuba.readthedocs.io/en/stable/src/api_reference.html#mitsuba.Mesh.vertex_texcoord

by the way, is it possible to update this method to use a texture for `specular` instead of a simple float? 
---------- previous reply---------------- 
my issue is I want to load a uv diffuse texture picture for `base_color`, using `bitmap` , to render a face mesh. But the uv texture seems doesn't match to mesh. I noticed `has_vertex_texcoords` in mi.mesh, but how can i add vertex_texcoords to it ? with `add_attribute`? or is there any other way to match uvtexture in `bsdf` to mesh?

uv texture is like:
![diff_tex](https://user-images.githubusercontent.com/23047054/208217219-0207276f-5fab-46fd-893d-5114ca56efe7.png)

**Reply by anewusername77** [reply]
> Any updates? I'm facing the same issue with using a specular map with `principled` BSDF

just use diffuse tex in base_color, and specular with a simple float

**Reply by giuvecchio** [reply]
My issue is with a bitmap ```specular``` map. Loading it into the ```specular_tint``` parameter doesn't seem to affect the rendering,

![maps](https://user-images.githubusercontent.com/19886419/208393914-2f457fe9-4352-4189-8f11-257085bbc4b1.jpg)

These are the maps used. What should the correct way be to use those?

**Reply by anewusername77** [reply]
> My issue is with a bitmap `specular` map. Loading it into the `specular_tint` parameter doesn't seem to affect the rendering,
> 
> ![maps](https://user-images.githubusercontent.com/19886419/208393914-2f457fe9-4352-4189-8f11-257085bbc4b1.jpg)
> 
> These are the maps used. What should the correct way be to use those?

yes my specular map was also a texture, but I don't think in`principled bsdf` specular can be a texture....so I just change it to a float

**Comment by NMontanaBrown** [comment]
Hiya, I have encountered the same issue and managed to resolve it the following way.

I have a .ply mesh with ST coordinates and it's corresponding baked diffuse map from Blender. Checking the scene parameters through params['vertex_texcoords'] as suggested above seems to return the same coords as in the .ply file.

The renderer expects UV coordinates though, so for conversion to UV:

`python

tex = params["PLYMesh.vertex_texcoords"]

new_tex = []
for i, item in enumerate(tex):
    if (i%2)==0:
        new_tex.append(item)
    else:
        new_tex.append(1-item)

params.update()`


The above applies the conversion u=s v=1-t.

I get the following renders with and without applying this change. Hope this helps anyone that comes across this!

![Figure_1_ST](https://user-images.githubusercontent.com/56116848/235501164-36bf5b32-d013-4697-b86b-a3e4282953f3.png)
![Figure_1_UV](https://user-images.githubusercontent.com/56116848/235501295-e92a7609-bd04-4dc9-9262-ff5ef12fe5b6.png)
